edgedb/protocol/codecs/namedtuple.pyx

```diff
@@ -70,15 +70,26 @@ cdef class NamedTupleCodec(BaseNamedRecordCodec):
         self._check_encoder()
 
         objlen = len(obj)
-        if objlen != len(self.fields_codecs):
+        fields_len = len(self.fields_codecs)
+
+        required_arguments = set()
+
+        for i in range(fields_len):
+            name = datatypes.record_desc_pointer_name(self.descriptor, i)
+            required_arguments.add(name)
+
+        if objlen != fields_len:
+            passed_arguments = set(obj.keys())
+            missed_arguments = required_arguments - passed_arguments
+            extra_arguments = passed_arguments - required_arguments
             raise RuntimeError(
-                f'expected {len(self.fields_codecs)} keyword arguments, '
-                f'got {objlen}')
+                f'expected {required_arguments} keyword arguments, '
+                f'got {passed_arguments}, missed {missed_arguments}, '
+                f'extra {extra_arguments}')
 
         elem_data = WriteBuffer.new()
-        for i in range(objlen):
-            name = datatypes.record_desc_pointer_name(self.descriptor, i)
-            arg = obj[name]
+        for i, required_arg in enumerate(required_arguments):
```

Comment:
You can't iterate over `required_arguments` here. Sets in Python are unordered and we must preserve the sequential order of arguments here. So please keep the old

```
for i in range(objlen):
    name = datatypes.record_desc_pointer_name(self.descriptor, i)
```

loop here.  What I suggest to do to actually print a nice error message in all cases is to refactor the code as follows:

```
for i in range(objlen):
    name = datatypes.record_desc_pointer_name(self.descriptor, i)
    try:
        arg = obj[name]
    except KeyError:
        raise self._make_missing_arg_error_message(obj) from None
```

and then you can implement the actual validation logic in the new `_make_missing_arg_error_message` method. That method can be rather complex as it will only be called when some argument is actually missing.

You should also keep the original `if objlen != len(self.fields_codecs):`, just change it to

```
if objlen != len(self.fields_codecs):
    raise raise self._make_missing_arg_error_message(obj)
```

Reply:
Thanks for your suggestions!

Moved the creation of the error and added new tests.

Reply:
Nice, please take a look at the one last nit comment.